File: internal/automation/profile_form_filler.go
package automation

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/ai-form-filler/cli/internal/models"
	"github.com/playwright-community/playwright-go"
)

// ProfileFormFiller handles form filling using client profiles with intelligent field mapping
type ProfileFormFiller struct {
	formFiller     *FormFiller
	formDetector   *FormDetector
	templateManager *TemplateManager
	fieldMapper    *FieldMapper
	config         *ProfileFormFillerConfig
}

// ProfileFormFillerConfig holds configuration for profile-based form filling
type ProfileFormFillerConfig struct {
	AutoDetectFields    bool
	UseAIMapping        bool
	VerifySubmission    bool
	MaxRetries          int
	RetryDelay          time.Duration
	SubmissionTimeout   time.Duration
	FieldMappingTimeout time.Duration
}

// DefaultProfileFormFillerConfig returns sensible defaults
func DefaultProfileFormFillerConfig() *ProfileFormFillerConfig {
	return &ProfileFormFillerConfig{
		AutoDetectFields:    true,
		UseAIMapping:        false, // Will be enabled when AI integration is added
		VerifySubmission:    true,
		MaxRetries:          3,
		RetryDelay:          2 * time.Second,
		SubmissionTimeout:   30 * time.Second,
		FieldMappingTimeout: 10 * time.Second,
	}
}

// ProfileFillResult represents the result of profile-based form filling
type ProfileFillResult struct {
	*FillResult
	ProfileID       string                 `json:"profileId"`
	TemplateUsed    *FormTemplate          `json:"templateUsed,omitempty"`
	FieldMappings   map[string]string      `json:"fieldMappings"`
	UnmappedFields  []string               `json:"unmappedFields"`
	SubmissionResult *SubmissionResult     `json:"submissionResult,omitempty"`
	Confidence      float64                `json:"confidence"`
}

// SubmissionResult represents the result of form submission
type SubmissionResult struct {
	Success         bool          `json:"success"`
	SubmissionTime  time.Duration `json:"submissionTime"`
	RedirectURL     string        `json:"redirectUrl,omitempty"`
	SuccessIndicators []string    `json:"successIndicators"`
	ErrorMessages   []string      `json:"errorMessages"`
	StatusCode      int           `json:"statusCode,omitempty"`
}

// NewProfileFormFiller creates a new profile-based form filler
func NewProfileFormFiller(
	formFiller *FormFiller,
	formDetector *FormDetector,
	templateManager *TemplateManager,
	config *ProfileFormFillerConfig,
) *ProfileFormFiller {
	if config == nil {
		config = DefaultProfileFormFillerConfig()
	}

	return &ProfileFormFiller{
		formFiller:      formFiller,
		formDetector:    formDetector,
		templateManager: templateManager,
		fieldMapper:     NewFieldMapper(),
		config:          config,
	}
}

// FillFormWithProfile fills a form using a client profile
func (pff *ProfileFormFiller) FillFormWithProfile(
	ctx context.Context,
	pageURL string,
	profile *models.ClientProfile,
) (*ProfileFillResult, error) {
	startTime := time.Now()

	// Try to find existing template first
	template, err := pff.templateManager.FindBestTemplate(pageURL)
	if err != nil && pff.config.AutoDetectFields {
		// No template found, analyze the page to create one
		analysis, err := pff.formDetector.AnalyzePage(ctx, pageURL)
		if err != nil {
			return nil, fmt.Errorf("failed to analyze page: %w", err)
		}

		if len(analysis.Forms) == 0 {
			return nil, fmt.Errorf("no forms detected on page: %s", pageURL)
		}

		// Use the form with highest confidence
		bestForm := analysis.Forms[0]
		template, err = pff.formDetector.GenerateFormTemplate(bestForm, pageURL)
		if err != nil {
			return nil, fmt.Errorf("failed to generate template: %w", err)
		}

		// Save the new template
		if err := pff.templateManager.SaveTemplate(template); err != nil {
			// Log error but continue - template saving is not critical for filling
			fmt.Printf("Warning: failed to save template: %v\n", err)
		}
	} else if err != nil {
		return nil, fmt.Errorf("no template found and auto-detection disabled: %w", err)
	}

	// Map profile data to form fields
	fieldMappings, unmappedFields := pff.fieldMapper.MapProfileToFields(profile, template.Fields)

	// Create ProfileData for the form filler
	profileData := pff.convertToProfileData(profile)

	// Fill the form
	fillResult, err := pff.formFiller.FillForm(ctx, template, profileData)
	if err != nil {
		return nil, fmt.Errorf("failed to fill form: %w", err)
	}

	// Calculate confidence based on mapping success
	confidence := pff.calculateMappingConfidence(fieldMappings, template.Fields)

	result := &ProfileFillResult{
		FillResult:     fillResult,
		ProfileID:      profile.ID,
		TemplateUsed:   template,
		FieldMappings:  fieldMappings,
		UnmappedFields: unmappedFields,
		Confidence:     confidence,
	}

	// Submit form if configured to do so
	if pff.config.VerifySubmission && fillResult.Success {
		submissionResult, err := pff.submitAndVerify(ctx, template, pageURL)
		if err != nil {
			// Log error but don't fail the entire operation
			fmt.Printf("Warning: form submission failed: %v\n", err)
		} else {
			result.SubmissionResult = submissionResult
		}
	}

	// Update template success rate
	successRate := 0.0
	if fillResult.Success {
		successRate = fillResult.SuccessRate
		if result.SubmissionResult != nil && result.SubmissionResult.Success {
			successRate = 100.0
		}
	}

	if err := pff.templateManager.UpdateTemplateSuccess(template.ID, successRate); err != nil {
		fmt.Printf("Warning: failed to update template success rate: %v\n", err)
	}

	result.ExecutionTime = time.Since(startTime)
	return result, nil
}

// convertToProfileData converts a ClientProfile to ProfileData for the form filler
func (pff *ProfileFormFiller) convertToProfileData(profile *models.ClientProfile) *ProfileData {
	return &ProfileData{
		FirstName: profile.PersonalData.FirstName,
		LastName:  profile.PersonalData.LastName,
		Email:     profile.PersonalData.Email,
		Phone:     profile.PersonalData.Phone,
		Address: Address{
			Street:  profile.PersonalData.Address.Street1,
			City:    profile.PersonalData.Address.City,
			State:   profile.PersonalData.Address.State,
			ZipCode: profile.PersonalData.Address.PostalCode,
			Country: profile.PersonalData.Address.Country,
		},
	}
}

// calculateMappingConfidence calculates confidence based on how many fields were successfully mapped
func (pff *ProfileFormFiller) calculateMappingConfidence(mappings map[string]string, fields []FormField) float64 {
	if len(fields) == 0 {
		return 0.0
	}

	mappedCount := 0
	for _, field := range fields {
		if _, mapped := mappings[field.Name]; mapped {
			mappedCount++
		}
	}

	return float64(mappedCount) / float64(len(fields)) * 100.0
}

// submitAndVerify submits the form and verifies successful submission
func (pff *ProfileFormFiller) submitAndVerify(ctx context.Context, template *FormTemplate, pageURL string) (*SubmissionResult, error) {
	// Create a new page for submission
	page, err := pff.formFiller.browserManager.CreatePage(DefaultBrowserConfig())
	if err != nil {
		return nil, fmt.Errorf("failed to create page for submission: %w", err)
	}
	defer (*page).Close()

	// Navigate to the page
	_, err = (*page).Goto(pageURL, playwright.PageGotoOptions{
		WaitUntil: playwright.WaitUntilStateNetworkidle,
		Timeout:   playwright.Float(float64(pff.config.SubmissionTimeout.Milliseconds())),
	})
	if err != nil {
		return nil, fmt.Errorf("failed to navigate to page: %w", err)
	}

	startTime := time.Now()

	// Submit the form
	err = pff.formFiller.SubmitForm(ctx, page, template)
	if err != nil {
		return &SubmissionResult{
			Success:        false,
			SubmissionTime: time.Since(startTime),
			ErrorMessages:  []string{err.Error()},
		}, nil
	}

	// Wait for navigation or response
	time.Sleep(2 * time.Second)

	// Get current URL to check for redirect
	currentURL := (*page).URL()

	// Check for success indicators
	successIndicators, errorMessages := pff.checkSubmissionResult(ctx, page)

	success := len(errorMessages) == 0 && (currentURL != pageURL || len(successIndicators) > 0)

	return &SubmissionResult{
		Success:           success,
		SubmissionTime:    time.Since(startTime),
		RedirectURL:       currentURL,
		SuccessIndicators: successIndicators,
		ErrorMessages:     errorMessages,
	}, nil
}

// checkSubmissionResult checks the page for success or error indicators
func (pff *ProfileFormFiller) checkSubmissionResult(ctx context.Context, page *playwright.Page) ([]string, []string) {
	var successIndicators []string
	var errorMessages []string

	// Common success indicators
	successSelectors := []string{
		".success", ".alert-success", ".message-success",
		"[class*='success']", "[id*='success']",
		".confirmation", ".thank-you", ".complete",
	}

	// Common error indicators
	errorSelectors := []string{
		".error", ".alert-error", ".message-error",
		"[class*='error']", "[id*='error']",
		".warning", ".alert-warning", ".invalid",
	}

	// Check for success indicators
	for _, selector := range successSelectors {
		elements, err := (*page).QuerySelectorAll(selector)
		if err != nil {
			continue
		}

		for _, element := range elements {
			text, err := element.TextContent()
			if err != nil {
				continue
			}
			if text != "" && strings.TrimSpace(text) != "" {
				successIndicators = append(successIndicators, strings.TrimSpace(text))
			}
		}
	}

	// Check for error indicators
	for _, selector := range errorSelectors {
		elements, err := (*page).QuerySelectorAll(selector)
		if err != nil {
			continue
		}

		for _, element := range elements {
			text, err := element.TextContent()
			if err != nil {
				continue
			}
			if text != "" && strings.TrimSpace(text) != "" {
				errorMessages = append(errorMessages, strings.TrimSpace(text))
			}
		}
	}

	// Check page title for success/error keywords
	title, err := (*page).Title()
	if err == nil {
		lowerTitle := strings.ToLower(title)
		if strings.Contains(lowerTitle, "success") || strings.Contains(lowerTitle, "thank") || strings.Contains(lowerTitle, "complete") {
			successIndicators = append(successIndicators, fmt.Sprintf("Title: %s", title))
		}
		if strings.Contains(lowerTitle, "error") || strings.Contains(lowerTitle, "failed") {
			errorMessages = append(errorMessages, fmt.Sprintf("Title: %s", title))
		}
	}

	return successIndicators, errorMessages
}

// FieldMapper handles mapping between profile data and form fields
type FieldMapper struct {
	fieldPatterns map[string][]*regexp.Regexp
}

// NewFieldMapper creates a new field mapper with predefined patterns
func NewFieldMapper() *FieldMapper {
	fm := &FieldMapper{
		fieldPatterns: make(map[string][]*regexp.Regexp),
	}

	// Initialize field mapping patterns
	fm.initializePatterns()
	return fm
}

// initializePatterns sets up regex patterns for field mapping
func (fm *FieldMapper) initializePatterns() {
	patterns := map[string][]string{
		"firstName": {
			`(?i)first.*name`, `(?i)fname`, `(?i)given.*name`,
			`(?i)forename`, `(?i)prenom`, `(?i)nombre`,
		},
		"lastName": {
			`(?i)last.*name`, `(?i)lname`, `(?i)surname`,
			`(?i)family.*name`, `(?i)apellido`, `(?i)nom`,
		},
		"email": {
			`(?i)email`, `(?i)e.mail`, `(?i)mail`,
			`(?i)correo`, `(?i)courriel`,
		},
		"phone": {
			`(?i)phone`, `(?i)tel`, `(?i)mobile`, `(?i)cell`,
			`(?i)telefono`, `(?i)telephone`, `(?i)numero`,
		},
		"address": {
			`(?i)address`, `(?i)street`, `(?i)addr`,
			`(?i)direccion`, `(?i)adresse`, `(?i)rue`,
		},
		"city": {
			`(?i)city`, `(?i)town`, `(?i)ciudad`,
			`(?i)ville`, `(?i)locality`,
		},
		"state": {
			`(?i)state`, `(?i)province`, `(?i)region`,
			`(?i)estado`, `(?i)provincia`, `(?i)departement`,
		},
		"zipCode": {
			`(?i)zip`, `(?i)postal`, `(?i)postcode`,
			`(?i)codigo.*postal`, `(?i)code.*postal`,
		},
		"country": {
			`(?i)country`, `(?i)nation`, `(?i)pais`,
			`(?i)pays`, `(?i)nationality`,
		},
	}

	for fieldType, patternStrings := range patterns {
		var regexps []*regexp.Regexp
		for _, pattern := range patternStrings {
			if regex, err := regexp.Compile(pattern); err == nil {
				regexps = append(regexps, regex)
			}
		}
		fm.fieldPatterns[fieldType] = regexps
	}
}

// MapProfileToFields maps profile data to form fields based on field names and labels
func (fm *FieldMapper) MapProfileToFields(profile *models.ClientProfile, fields []FormField) (map[string]string, []string) {
	mappings := make(map[string]string)
	var unmappedFields []string

	profileData := map[string]string{
		"firstName": profile.PersonalData.FirstName,
		"lastName":  profile.PersonalData.LastName,
		"email":     profile.PersonalData.Email,
		"phone":     profile.PersonalData.Phone,
		"address":   profile.PersonalData.Address.Street1,
		"city":      profile.PersonalData.Address.City,
		"state":     profile.PersonalData.Address.State,
		"zipCode":   profile.PersonalData.Address.PostalCode,
		"country":   profile.PersonalData.Address.Country,
	}

	for _, field := range fields {
		mapped := false

		// Try to map based on field name and label
		for profileField, value := range profileData {
			if value == "" {
				continue // Skip empty profile values
			}

			if fm.matchesField(field, profileField) {
				mappings[field.Name] = value
				mapped = true
				break
			}
		}

		// Handle special cases
		if !mapped {
			if specialValue := fm.handleSpecialField(field, profile); specialValue != "" {
				mappings[field.Name] = specialValue
				mapped = true
			}
		}

		if !mapped {
			unmappedFields = append(unmappedFields, field.Name)
		}
	}

	return mappings, unmappedFields
}

// matchesField checks if a form field matches a profile field type
func (fm *FieldMapper) matchesField(field FormField, profileFieldType string) bool {
	patterns, exists := fm.fieldPatterns[profileFieldType]
	if !exists {
		return false
	}

	// Check field name
	for _, pattern := range patterns {
		if pattern.MatchString(field.Name) {
			return true
		}
	}

	// Check field label
	for _, pattern := range patterns {
		if pattern.MatchString(field.Label) {
			return true
		}
	}

	// Check field type for email and tel
	if profileFieldType == "email" && field.Type == "email" {
		return true
	}
	if profileFieldType == "phone" && (field.Type == "tel" || field.Type == "phone") {
		return true
	}

	return false
}

// handleSpecialField handles special field types that need custom logic
func (fm *FieldMapper) handleSpecialField(field FormField, profile *models.ClientProfile) string {
	fieldName := strings.ToLower(field.Name)
	fieldLabel := strings.ToLower(field.Label)

	// Handle full name fields
	if strings.Contains(fieldName, "fullname") || strings.Contains(fieldLabel, "full name") {
		return fmt.Sprintf("%s %s", profile.PersonalData.FirstName, profile.PersonalData.LastName)
	}

	// Handle name fields (could be first or last)
	if fieldName == "name" || fieldLabel == "name" {
		// If it's the only name field, use full name
		return fmt.Sprintf("%s %s", profile.PersonalData.FirstName, profile.PersonalData.LastName)
	}

	// Handle date of birth
	if strings.Contains(fieldName, "birth") || strings.Contains(fieldLabel, "birth") ||
		strings.Contains(fieldName, "dob") || strings.Contains(fieldLabel, "date of birth") {
		return profile.PersonalData.DateOfBirth
	}

	// Handle custom fields
	for key, value := range profile.PersonalData.CustomFields {
		if strings.Contains(fieldName, strings.ToLower(key)) || strings.Contains(fieldLabel, strings.ToLower(key)) {
			if strValue, ok := value.(string); ok {
				return strValue
			}
		}
	}

	return ""
}

// ValidateFieldMapping validates that field mappings are appropriate
func (fm *FieldMapper) ValidateFieldMapping(mappings map[string]string, fields []FormField) []string {
	var warnings []string

	for _, field := range fields {
		if value, mapped := mappings[field.Name]; mapped {
			// Check required fields first
			if field.Required && strings.TrimSpace(value) == "" {
				warnings = append(warnings, fmt.Sprintf("Required field %s is empty", field.Name))
				continue
			}

			// Skip validation for empty non-required fields
			if strings.TrimSpace(value) == "" {
				continue
			}

			// Validate email format
			if field.Type == "email" {
				emailRegex := regexp.MustCompile(`^[^@\s]+@[^@\s]+\.[^@\s]+$`)
				if !emailRegex.MatchString(value) {
					warnings = append(warnings, fmt.Sprintf("Invalid email format for field %s: %s", field.Name, value))
				}
			}

			// Validate phone format (basic check)
			if field.Type == "tel" || field.Type == "phone" {
				phoneRegex := regexp.MustCompile(`^[\d\s\-\(\)\+]+$`)
				if !phoneRegex.MatchString(value) {
					warnings = append(warnings, fmt.Sprintf("Invalid phone format for field %s: %s", field.Name, value))
				}
			}
		} else if field.Required {
			warnings = append(warnings, fmt.Sprintf("Required field %s is not mapped", field.Name))
		}
	}

	return warnings
}

File: archive/backup-2025-01-14/shared-package/src/profile-form-filler.ts
import { ClientProfile, FormField, FormTemplate } from './types';
import { FormDetector, DetectedForm, DetectedField } from './form-detection';

/**
 * Profile-based form filler for intelligent form completion using client profiles
 */

export interface ProfileFormFillerConfig {
  autoDetectFields: boolean;
  useAIMapping: boolean;
  verifySubmission: boolean;
  maxRetries: number;
  retryDelay: number; // milliseconds
  submissionTimeout: number; // milliseconds
  fieldMappingTimeout: number; // milliseconds
}

export interface ProfileFillResult {
  success: boolean;
  profileId: string;
  templateUsed?: FormTemplate;
  fieldMappings: Record<string, string>;
  unmappedFields: string[];
  submissionResult?: SubmissionResult;
  confidence: number;
  filledFields: number;
  totalFields: number;
  executionTime: number;
  errors: string[];
}

export interface SubmissionResult {
  success: boolean;
  submissionTime: number;
  redirectUrl?: string;
  successIndicators: string[];
  errorMessages: string[];
  statusCode?: number;
}

export interface FieldMapping {
  fieldName: string;
  profileField: string;
  value: string;
  confidence: number;
}

export class ProfileFormFiller {
  private formDetector: FormDetector;
  private fieldMapper: FieldMapper;
  private config: ProfileFormFillerConfig;

  constructor(config?: Partial<ProfileFormFillerConfig>) {
    this.formDetector = new FormDetector();
    this.fieldMapper = new FieldMapper();
    this.config = {
      autoDetectFields: true,
      useAIMapping: false,
      verifySubmission: true,
      maxRetries: 3,
      retryDelay: 2000,
      submissionTimeout: 30000,
      fieldMappingTimeout: 10000,
      ...config
    };
  }

  /**
   * Fills a form using a client profile
   */
  public async fillFormWithProfile(
    profile: ClientProfile,
    template?: FormTemplate
  ): Promise<ProfileFillResult> {
    const startTime = performance.now();

    try {
      // If no template provided, detect forms on current page
      let formTemplate = template;
      if (!formTemplate && this.config.autoDetectFields) {
        const analysis = this.formDetector.analyzePage();
        if (analysis.forms.length === 0) {
          throw new Error('No forms detected on page');
        }

        // Use the form with highest confidence
        const bestForm = analysis.forms[0];
        formTemplate = this.generateTemplateFromDetectedForm(bestForm);
      }

      if (!formTemplate) {
        throw new Error('No template available and auto-detection disabled');
      }

      // Map profile data to form fields
      const { mappings, unmappedFields } = this.fieldMapper.mapProfileToFields(
        profile,
        formTemplate.fields
      );

      // Fill the form
      const fillResult = await this.fillFormFields(mappings, formTemplate);

      // Calculate confidence
      const confidence = this.calculateMappingConfidence(mappings, formTemplate.fields);

      const result: ProfileFillResult = {
        success: fillResult.success,
        profileId: profile.id,
        templateUsed: formTemplate,
        fieldMappings: mappings,
        unmappedFields,
        confidence,
        filledFields: fillResult.filledFields,
        totalFields: formTemplate.fields.length,
        executionTime: performance.now() - startTime,
        errors: fillResult.errors
      };

      // Submit form if configured to do so
      if (this.config.verifySubmission && fillResult.success) {
        try {
          const submissionResult = await this.submitAndVerify(formTemplate);
          result.submissionResult = submissionResult;
        } catch (error) {
          result.errors.push(`Submission failed: ${error}`);
        }
      }

      return result;
    } catch (error) {
      return {
        success: false,
        profileId: profile.id,
        fieldMappings: {},
        unmappedFields: [],
        confidence: 0,
        filledFields: 0,
        totalFields: 0,
        executionTime: performance.now() - startTime,
        errors: [error instanceof Error ? error.message : String(error)]
      };
    }
  }

  /**
   * Generates a template from a detected form
   */
  private generateTemplateFromDetectedForm(detectedForm: DetectedForm): FormTemplate {
    const url = window.location.href;
    const domain = window.location.hostname;

    return {
      id: `template_${domain.replace(/\./g, '_')}_${detectedForm.formType}_${Date.now()}`,
      url,
      domain,
      formType: detectedForm.formType,
      fields: detectedForm.fields.map((field: DetectedField) => ({
        id: `field_${field.name}_${Date.now()}`,
        name: field.name,
        type: field.type as any,
        selector: field.selector,
        label: field.label,
        required: field.required,
        validationPattern: field.validationPattern,
        defaultValue: field.placeholder
      })),
      selectors: detectedForm.fields.reduce((acc: Record<string, string>, field: DetectedField) => {
        acc[field.name] = field.selector;
        return acc;
      }, {} as Record<string, string>),
      validationRules: [],
      successRate: 0,
      lastUpdated: new Date(),
      version: 1
    };
  }

  /**
   * Fills form fields with mapped values
   */
  private async fillFormFields(
    mappings: Record<string, string>,
    template: FormTemplate
  ): Promise<{ success: boolean; filledFields: number; errors: string[] }> {
    const errors: string[] = [];
    let filledFields = 0;

    for (const field of template.fields) {
      const value = mappings[field.name];
      if (!value) {
        continue;
      }

      try {
        const element = document.querySelector(field.selector);
        if (!element) {
          errors.push(`Element not found for selector: ${field.selector}`);
          continue;
        }

        await this.fillElement(element as HTMLElement, value, field.type);
        filledFields++;

        // Add small delay between field fills
        await this.delay(100);
      } catch (error) {
        errors.push(`Failed to fill field ${field.name}: ${error}`);
      }
    }

    return {
      success: filledFields > 0 && errors.length === 0,
      filledFields,
      errors
    };
  }

  /**
   * Fills a single form element
   */
  private async fillElement(element: HTMLElement, value: string, fieldType: string): Promise<void> {
    if (element instanceof HTMLInputElement) {
      switch (fieldType) {
        case 'checkbox':
          element.checked = value === 'true' || value === '1' || value.toLowerCase() === 'yes';
          break;
        case 'radio':
          if (element.value === value) {
            element.checked = true;
          }
          break;
        default:
          element.value = value;
          break;
      }
    } else if (element instanceof HTMLSelectElement) {
      // Try to select by value first, then by text
      const option = Array.from(element.options).find(
        opt => opt.value === value || opt.text === value
      );
      if (option) {
        element.selectedIndex = option.index;
      }
    } else if (element instanceof HTMLTextAreaElement) {
      element.value = value;
    }

    // Trigger events to notify the page of changes
    element.dispatchEvent(new Event('input', { bubbles: true }));
    element.dispatchEvent(new Event('change', { bubbles: true }));
  }

  /**
   * Submits the form and verifies successful submission
   */
  private async submitAndVerify(template: FormTemplate): Promise<SubmissionResult> {
    const startTime = performance.now();
    const currentUrl = window.location.href;

    // Find and click submit button
    const submitButton = this.findSubmitButton();
    if (!submitButton) {
      throw new Error('No submit button found');
    }

    submitButton.click();

    // Wait for navigation or response
    await this.delay(2000);

    // Check for success/error indicators
    const { successIndicators, errorMessages } = this.checkSubmissionResult();

    const newUrl = window.location.href;
    const success = errorMessages.length === 0 && (newUrl !== currentUrl || successIndicators.length > 0);

    return {
      success,
      submissionTime: performance.now() - startTime,
      redirectUrl: newUrl !== currentUrl ? newUrl : undefined,
      successIndicators,
      errorMessages
    };
  }

  /**
   * Finds the submit button on the page
   */
  private findSubmitButton(): HTMLElement | null {
    const selectors = [
      'input[type="submit"]',
      'button[type="submit"]',
      'button:not([type])', // buttons default to submit
      'button:contains("Submit")',
      'button:contains("Send")',
      'button:contains("Continue")',
      'input[value*="Submit" i]'
    ];

    for (const selector of selectors) {
      const element = document.querySelector(selector);
      if (element) {
        return element as HTMLElement;
      }
    }

    return null;
  }

  /**
   * Checks the page for success or error indicators
   */
  private checkSubmissionResult(): { successIndicators: string[]; errorMessages: string[] } {
    const successIndicators: string[] = [];
    const errorMessages: string[] = [];

    // Success selectors
    const successSelectors = [
      '.success', '.alert-success', '.message-success',
      '[class*="success"]', '[id*="success"]',
      '.confirmation', '.thank-you', '.complete'
    ];

    // Error selectors
    const errorSelectors = [
      '.error', '.alert-error', '.message-error',
      '[class*="error"]', '[id*="error"]',
      '.warning', '.alert-warning', '.invalid'
    ];

    // Check for success indicators
    successSelectors.forEach(selector => {
      const elements = document.querySelectorAll(selector);
      elements.forEach(element => {
        const text = element.textContent?.trim();
        if (text) {
          successIndicators.push(text);
        }
      });
    });

    // Check for error indicators
    errorSelectors.forEach(selector => {
      const elements = document.querySelectorAll(selector);
      elements.forEach(element => {
        const text = element.textContent?.trim();
        if (text) {
          errorMessages.push(text);
        }
      });
    });

    // Check page title
    const title = document.title.toLowerCase();
    if (title.includes('success') || title.includes('thank') || title.includes('complete')) {
      successIndicators.push(`Title: ${document.title}`);
    }
    if (title.includes('error') || title.includes('failed')) {
      errorMessages.push(`Title: ${document.title}`);
    }

    return { successIndicators, errorMessages };
  }

  /**
   * Calculates confidence based on mapping success
   */
  private calculateMappingConfidence(mappings: Record<string, string>, fields: FormField[]): number {
    if (fields.length === 0) return 0;

    const mappedCount = fields.filter(field => mappings[field.name]).length;
    return (mappedCount / fields.length) * 100;
  }

  /**
   * Utility delay function
   */
  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

/**
 * Field mapper for mapping profile data to form fields
 */
export class FieldMapper {
  private fieldPatterns: Record<string, RegExp[]>;

  constructor() {
    this.fieldPatterns = this.initializePatterns();
  }

  /**
   * Maps profile data to form fields
   */
  public mapProfileToFields(
    profile: ClientProfile,
    fields: FormField[]
  ): { mappings: Record<string, string>; unmappedFields: string[] } {
    const mappings: Record<string, string> = {};
    const unmappedFields: string[] = [];

    const profileData = this.extractProfileData(profile);

    for (const field of fields) {
      let mapped = false;

      // Try to map based on field name and label
      for (const [profileField, value] of Object.entries(profileData)) {
        if (!value) continue; // Skip empty values

        if (this.matchesField(field, profileField)) {
          mappings[field.name] = value;
          mapped = true;
          break;
        }
      }

      // Handle special cases
      if (!mapped) {
        const specialValue = this.handleSpecialField(field, profile);
        if (specialValue) {
          mappings[field.name] = specialValue;
          mapped = true;
        }
      }

      if (!mapped) {
        unmappedFields.push(field.name);
      }
    }

    return { mappings, unmappedFields };
  }

  /**
   * Extracts profile data into a flat structure
   */
  private extractProfileData(profile: ClientProfile): Record<string, string> {
    return {
      firstName: profile.personalData.firstName,
      lastName: profile.personalData.lastName,
      email: profile.personalData.email,
      phone: profile.personalData.phone || '',
      address: profile.personalData.address?.street || '',
      city: profile.personalData.address?.city || '',
      state: profile.personalData.address?.state || '',
      zipCode: profile.personalData.address?.zipCode || '',
      country: profile.personalData.address?.country || ''
    };
  }

  /**
   * Checks if a form field matches a profile field type
   */
  private matchesField(field: FormField, profileFieldType: string): boolean {
    const patterns = this.fieldPatterns[profileFieldType];
    if (!patterns) return false;

    // Check field name
    for (const pattern of patterns) {
      if (pattern.test(field.name)) return true;
    }

    // Check field label
    for (const pattern of patterns) {
      if (pattern.test(field.label)) return true;
    }

    // Check field type for email and tel
    if (profileFieldType === 'email' && field.type === 'email') return true;
    if (profileFieldType === 'phone' && (field.type === 'tel' || field.type === 'phone')) return true;

    return false;
  }

  /**
   * Handles special field types that need custom logic
   */
  private handleSpecialField(field: FormField, profile: ClientProfile): string {
    const fieldName = field.name.toLowerCase();
    const fieldLabel = field.label.toLowerCase();

    // Handle full name fields
    if (fieldName.includes('fullname') || fieldLabel.includes('full name')) {
      return `${profile.personalData.firstName} ${profile.personalData.lastName}`;
    }

    // Handle generic name fields
    if (fieldName === 'name' || fieldLabel === 'name') {
      return `${profile.personalData.firstName} ${profile.personalData.lastName}`;
    }

    // Handle date of birth
    if (fieldName.includes('birth') || fieldLabel.includes('birth') ||
        fieldName.includes('dob') || fieldLabel.includes('date of birth')) {
      return profile.personalData.dateOfBirth?.toISOString().split('T')[0] || '';
    }

    // Handle custom fields
    if (profile.personalData.customFields) {
      for (const [key, value] of Object.entries(profile.personalData.customFields)) {
        if (fieldName.includes(key.toLowerCase()) || fieldLabel.includes(key.toLowerCase())) {
          return String(value);
        }
      }
    }

    return '';
  }

  /**
   * Initializes field mapping patterns
   */
  private initializePatterns(): Record<string, RegExp[]> {
    const patterns: Record<string, string[]> = {
      firstName: [
        '^first.*name$', '^fname$', '^given.*name$',
        '^forename$', '^prenom$', '^nombre$'
      ],
      lastName: [
        '^last.*name$', '^lname$', '^surname$',
        '^family.*name$', '^apellido$', '^nom$'
      ],
      email: [
        'email', 'e.mail', 'mail',
        'correo', 'courriel'
      ],
      phone: [
        'phone', 'tel', 'mobile', 'cell',
        'telefono', 'telephone', 'numero'
      ],
      address: [
        'address', 'street', 'addr',
        'direccion', 'adresse', 'rue'
      ],
      city: [
        'city', 'town', 'ciudad',
        'ville', 'locality'
      ],
      state: [
        'state', 'province', 'region',
        'estado', 'provincia', 'departement'
      ],
      zipCode: [
        'zip', 'postal', 'postcode',
        'codigo.*postal', 'code.*postal'
      ],
      country: [
        'country', 'nation', 'pais',
        'pays', 'nationality'
      ]
    };

    const regexPatterns: Record<string, RegExp[]> = {};
    for (const [fieldType, patternStrings] of Object.entries(patterns)) {
      regexPatterns[fieldType] = patternStrings.map(pattern => new RegExp(pattern, 'i'));
    }

    return regexPatterns;
  }

  /**
   * Validates field mappings
   */
  public validateFieldMapping(mappings: Record<string, string>, fields: FormField[]): string[] {
    const warnings: string[] = [];

    for (const field of fields) {
      const value = mappings[field.name];
      if (value === undefined) {
        if (field.required) {
          warnings.push(`Required field ${field.name} is not mapped`);
        }
        continue;
      }

      // Check required fields for empty values
      if (field.required && value.trim() === '') {
        warnings.push(`Required field ${field.name} is empty`);
        continue;
      }

      // Validate email format
      if (field.type === 'email' && value.trim() !== '') {
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (!emailRegex.test(value)) {
          warnings.push(`Invalid email format for field ${field.name}: ${value}`);
        }
      }

      // Validate phone format
      if ((field.type === 'tel' || field.type === 'phone') && value.trim() !== '') {
        const phoneRegex = /^[\d\s\-\(\)\+]+$/;
        if (!phoneRegex.test(value)) {
          warnings.push(`Invalid phone format for field ${field.name}: ${value}`);
        }
      }
    }

    return warnings;
  }
}